dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackedEntityInstanceService.java

```diff
@@ -757,16 +757,20 @@ public void validateSearchScope( TrackedEntityInstanceQueryParams params, boolea
                 }
             }
 
-            if ( maxTeiLimit > 0 && params.isPaging()
-                && params.getOffset() > 0 && (params.getOffset() + params.getPageSizeWithDefault()) > maxTeiLimit )
-            {
-                throw new IllegalQueryException( "maxteicountreached" );
-            }
-
+            checkIfMaxTeiLimitIsReached( params, maxTeiLimit );
             params.setMaxTeiLimit( maxTeiLimit );
         }
     }
 
+    private void checkIfMaxTeiLimitIsReached( TrackedEntityInstanceQueryParams params, int maxTeiLimit )
+    {
+        int instanceCount = trackedEntityInstanceStore.getTrackedEntityInstanceCountForGrid( params );
```

Comment:
The previous implementation didn't seem correct at all but this new one can have a big impact on performance.
We need to be sure that performance are not impacted too much by this change.

For sure one thing that can already be improved is to call the count only if maxTeiLimit > 0

Reply:
Good idea to run the query only if maxTeiLimit > 0. Other than that, I added a clause to limit the count to maxTeiLimit+1, I guess this will improve the performance a bit. Any other suggestion is welcome.

Reply:
Yeah, doing the limit+1 is probably the most efficient way to do this. I think taking into account pagesize actually allowed for circumventing; Ignoring paging looks like the best way to go